gapis/api/vulkan/vulkan.go

```diff
@@ -78,6 +78,137 @@ func (s *State) SetupInitialState(ctx context.Context, state *api.GlobalState) {
 	}
 }
 
+// TrimInitialState scans the capture commands to see which parts of the initial
+// state are actually used, and removes some unused parts from it.
+//
+// Note: the current approach consists in "manually" monitoring which Vulkan
+// objects are being used in callbacks passed to sync.MutateWithSubcommands,
+// however this basically re-encode some state tracking logic found in the API
+// files. A better way would be to use an api.StateWatcher and rely on api.RefID
+// to track which objects are accessed: this would avoid to re-encode state
+// tracking logic here. There might be some pitfalls though, e.g. when a command
+// just reads the handle of an object, the state watcher would not mark an
+// access to that object. For instance, when creating a derivate pipeline, a
+// VkPipeline handle is used in BasePipelineHandle, the API implementation reads
+// this handle, but does not access the corresponding object. So using
+// api.StateWatcher might need some wider design considerations.
+func (s *State) TrimInitialState(ctx context.Context, capturePath *path.Capture) error {
+	// Parts of the state we want to record the usage of.
+	descriptorSets := map[VkDescriptorSet]struct{}{}
+	pipelines := map[VkPipeline]struct{}{}
+
+	// Record usage in initial state.
+	for _, i := range s.LastComputeInfos().All() {
+		pipelines[i.ComputePipeline().VulkanHandle()] = struct{}{}
+	}
+	for _, i := range s.LastDrawInfos().All() {
+		for _, d := range i.DescriptorSets().All() {
+			descriptorSets[d.VulkanHandle()] = struct{}{}
+		}
+		pipelines[i.GraphicsPipeline().VulkanHandle()] = struct{}{}
+	}
+
+	// Record usage in the trace commands
+	// top-level commands
+	postCmdCb := func(s *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd) {
+		switch cmd := cmd.(type) {
+		case *VkFreeDescriptorSets:
+			count := cmd.DescriptorSetCount()
+			ds, err := cmd.PDescriptorSets().Slice(0, (uint64)(count), s.MemoryLayout).Read(ctx, cmd, s, nil)
+			if err != nil {
+				panic("Read error")
+			}
+			for _, d := range ds {
+				descriptorSets[d] = struct{}{}
+			}
+
+		case *VkDestroyPipeline:
+			pipelines[cmd.Pipeline()] = struct{}{}
+		}
+
+	}
+	// sub-commands
+	postSubCmdCb := func(state *api.GlobalState, subCmdIdx api.SubCmdIdx, cmd api.Cmd, i interface{}) {
+		vkState := GetState(state)
+		cmdRef, ok := i.(CommandReferenceʳ)
+		if !ok {
+			panic("In Vulkan, MutateWithSubcommands' postSubCmdCb 'interface{}' is not a CommandReferenceʳ")
```

Comment:
panic or error?

Reply:
This is really a "should never happen" thing, so panic is OK. We have the same pattern in the framegraph code.